ambry-network/src/main/java/com/github/ambry/network/NettyServerRequestResponseChannel.java

```diff
@@ -30,18 +36,34 @@
  */
 public class NettyServerRequestResponseChannel implements RequestResponseChannel {
   private static final Logger logger = LoggerFactory.getLogger(NettyServerRequestResponseChannel.class);
-  private final ArrayBlockingQueue<NetworkRequest> requestQueue;
   private final Http2ServerMetrics http2ServerMetrics;
+  private final NetworkRequestQueue networkRequestQueue;
+  private final Queue<NetworkRequest> unqueuedRequests = new ConcurrentLinkedQueue<>();
 
-  public NettyServerRequestResponseChannel(int queueSize, Http2ServerMetrics http2ServerMetrics) {
-    requestQueue = new ArrayBlockingQueue<>(queueSize);
+  public NettyServerRequestResponseChannel(NetworkConfig config, Http2ServerMetrics http2ServerMetrics) {
+    switch (config.requestQueueType) {
+      case ADAPTIVE_LIFO_CO_DEL:
+        this.networkRequestQueue =
+            new AdaptiveLifoCoDelNetworkRequestQueue(config.queuedMaxRequests, config.adaptiveLifoQueueThreshold,
+                config.adaptiveLifoQueueCodelTargetDelayMs, config.requestQueueTimeoutMs, SystemTime.getInstance());
+        break;
+      case BASIC_FIFO:
+        this.networkRequestQueue = new FifoNetworkRequestQueue(config.queuedMaxRequests, config.requestQueueTimeoutMs,
+            SystemTime.getInstance());
+        break;
+      default:
+        throw new IllegalArgumentException("Queue type not supported by channel: " + config.requestQueueType);
+    }
     this.http2ServerMetrics = http2ServerMetrics;
   }
 
   /** Send a request to be handled, potentially blocking until there is room in the queue for the request */
   @Override
   public void sendRequest(NetworkRequest request) throws InterruptedException {
-    requestQueue.put(request);
```

Comment:
Looks like we have two places to decide drop request.
1. NetworkRequestQueue has a blocking queue. It may drop request as well.
2. RequestResponseChannel has a unqueuedRequests queue if NetworkRequestQueue blocking queue is full. 

Is that possible to let RequestResponseChannel manage all the requests and decide the drop policy? Don't split it to two places. NetworkRequestQueue can have its own queue instead of the blocking queue.

Reply:
Yeah, true. Like you said, it is probably good to have `unqueuedRequests` queue as well in `NetworkRequestQueue`. But, with the latest commit, since I am fetching unqueued requests separately from `RequestResponseChannel`, I thought I will leave the `unqueuedRequests` queue in RequestResponseChannel. Please let me know what you think.